### PR TITLE
TablePanel: Adding sort order persistance

### LINKED ISSUE #23904: Tabel panel: Create features overview dashboard

### LINKED ISSUE #24522: TablePanel: Sort order not persisted
Work was started on this but not completed:

https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Table/Table.tsx#L54

### LINKED ISSUE #24645: 7.0.0-beta 3: Table panel "un-sorts" when refreshed
**What happened**:
Added a table panel to my dashboard.  Selected a column to sort on.  The ^ shows up correctly, and the table is sorted by that field, as I expect.  I set the dashboard auto-refresh to 5 seconds.  When the dashboard refreshes, the sorting is lost - it now sorts however it sorted when initially adding the panel.  Every time the dashboard refreshes, the sort is lost.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.0.0-beta3
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by torkelo**
This PR adds sort order persistence to new table panel and new gdev test dashboard for some of the table panel features. 

Fixes #24522. (Table sort persistence)
Fixes #24645 (Table sort resets on refresh)
Fixes #23904.  (Table features overview dashboard)

Afraid a lot will complain about the lack of persistence in the new table panel, but it's not critical for 7.0. But could be considered for patch release based on level of feedback on the lack of this feature in 7.0

Screen Shot 2020-05-14 at 19 46 45

### CODE REVIEW

**Comment by dprokop** [comment]
First impression - great!

**Comment by dprokop** [comment]
Found a problem - when you change field display name or add display name override, the sort is lost :( we need to update the sortBy option accordingly after the name change

File: packages/grafana-ui/src/components/Table/types.ts [hunk 1]
```diff
@@ -26,11 +26,11 @@ export interface TableRow {
 }
 
 export type TableFilterActionCallback = (key: string, value: string) => void;
-export type TableColumnResizeActionCallback = (fieldIndex: number, width: number) => void;
+export type TableColumnResizeActionCallback = (fieldDisplayName: string, width: number) => void;
 export type TableSortByActionCallback = (state: TableSortByFieldState[]) => void;
 
 export interface TableSortByFieldState {
-  fieldIndex: number;
+  displayName: string;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Table/types.ts:33]: Was thinking if actually Field having a guid isn't a good idea long term. For now displayName is OK but taking field options and overrides that will require us to update the sortBy property when they change doesn't sound optimal.

**Review by dprokop — APPROVED** [review]
@torkelo tested it and it works great. Not sure we should push it to 7.0, as I'm not 100% convinced about the displayName being the right identity for the sort order persistence. But since this is probably a bigger discussion, this might be a good enough solution.

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 2]
```diff
@@ -20,21 +14,10 @@ export class TablePanel extends Component<Props> {
     super(props);
   }
 
-  onColumnResize = (fieldIndex: number, width: number) => {
-    const { fieldConfig, data } = this.props;
+  onColumnResize = (fieldDisplayName: string, width: number) => {
```
Comment (ryantxu) [inline public/app/plugins/panel/table/TablePanel.tsx:17]: name may not be unique -- we should never use it as a key

File: packages/grafana-ui/src/components/Table/types.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/Table/types.ts:33]: how would we assign a guid?? needs to be matched to a query result. field display name is all we have.

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:17]: Do we have an alternative?

Comment (ryantxu) [reply public/app/plugins/panel/table/TablePanel.tsx:17]: if it were just transient - the fieldIndex would be better... but if we persist the values then name makes more sense.  Either way they will each have weird behaviors for things with dupilcate names.

But the override configs UI would look better if they are by name, so I think you are right

Comment (mckn) [reply public/app/plugins/panel/table/TablePanel.tsx:17]: hmm, was thinking the same but at the same time we are using it as the identifier in transforms and on other places.

Is it possible to combine a natural ID in some way? Like fieldName + some kind of query identifier?

**Review by mckn — APPROVED** [review]
I think this look really good! Great work!

The worries that have been mentioned about the displayName as identifier is valid but I don't have enough knowledge to suggest a better solution. However, we persist it as the identifier in the transforms. So I guess if we find a better solution we still need to migrate the persisted data in multiple places so it feels better to merge this in the mean time and see how big of a problem this is.

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 2 above)
Comment (mckn) [reply public/app/plugins/panel/table/TablePanel.tsx:17]: Maybe I am totally offside now but would the displayName + query refId work? I am guessing that the refId is identifying the actual query and not the execution of the query.

**Pull request merged**